feat: Adding support for extraContainers and StatefuleSet deployment on v2.2.2

Nicolas-Peiffer opened the pull request:
This cancels and replace a previous [PR 87](https://github.com/twuni/docker-registry.helm/pull/87).

The purpose of these modifications is to add support for deploying the registry container as a `StatefuleSet` STS instead of a `Deployment`. This need several modifications on the templates, because the STS handles volume differently than the `Deployment`. A source of inspiration for this modification is the Grafana Helm Chart https://github.com/grafana/helm-charts/tree/main/charts/grafana.

And also adding support for `extraContainers` to be able to add sidecar containers to the registry container to support calling OCI syncing tools like [dregsy](https://github.com/xelalexv/dregsy) or [skopeo](https://github.com/containers/skopeo) to populate the OCI registry.

```yaml
# values.yaml for the `twuni/docker-registry.helm` helm chart
[...]
#extraContainers: []
# Additional containers to be added
extraContainers:
  ## Side car container for registry synchronization with dregsy and skopeo
  - name: dregsy
    image: docker.io/xelalex/dregsy
    imagePullPolicy: IfNotPresent
    command: ['dregsy', '-config=/home/dregsy/config/dregsy-config.yaml']
    resources:
      requests:
        cpu: 10m
        memory: 32Mi
    securityContext:
      allowPrivilegeEscalation: false
      privileged: false
      runAsNonRoot: true
      capabilities:
        drop:
        - all
    volumeMounts:
      - name: dregsy-config
        mountPath: /home/dregsy/config/
        readOnly: true
[...]
```

Signed-off-by: Nicolas-Peiffer <102670102+Nicolas-Peiffer@users.noreply.github.com>

## Release Notes

* Adding support for deploying the container registry using a Kubernetes `StatefuleSet` manifest object with dedicated per-sts PVC. Compatible with `podAntiAffinity`.
* Adding support for `extraContainers` to support calling OCI syncing tools like [dregsy](https://github.com/xelalexv/dregsy) or [skopeo](https://github.com/containers/skopeo) to populate the OCI registry.

Nicolas-Peiffer committed:
feat: Adding support for extraContainers and StatefuleSet deployment

The purpose of these modifications is to add support for deploying the registry container as a StatefuleSet STS instead of a Deployment. This need several modifications on the templates, because the STS handles volume differently than the Deployment. A source of instpiration for this modification is the Grafana Helm Chart https://github.com/grafana/helm-charts/tree/main/charts/grafana. Adding support for extraContainers to be able to add sidecar containers to the registry container.

Signed-off-by: Nicolas-Peiffer <102670102+Nicolas-Peiffer@users.noreply.github.com>

pracskai commented:
Our use case is to run the Docker registry in proxy (pull-through cache) mode to accelerate and reduce traffic for image pulls. The goal is to run multiple replicas for high availability, but without relying on shared RWX storage, which can be operationally challenging or unavailable in some Kubernetes environments.

By using a StatefulSet with separate PVCs (RWO) per replica, each registry instance maintains its own local cache. This setup is acceptable because:

- The registry is read-through from the remote (e.g., Docker Hub), so there's no strict need for data replication between instances.

- We accept eventual inconsistency in cache contents across replicas. The worst case is a cache miss, which simply results in a pull from the remote source.

- We avoid the complexity and fragility of RWX volumes while still getting basic HA via independent pods with independent caches.

So, while a multi-replica statefulset may not suit a write-heavy or push-mode registry, it works well for proxy caching, especially when consistency isn’t critical and storage availability is limited to RWO PVCs.

Nicolas-Peiffer commented:
By the way: I added support for `extraContainers` because I ran [dregsy](https://github.com/xelalexv/dregsy) as a side car container that was periodically syncing the content of the OCI registry instances in my cluster.

I added a couple of explanation in the PR main first comment.

Maybe I should have done 2 different PR since the `StatefuleSet` feature is different than the `extraContainers` feature. Feel free to separate my PR into 2. I will not do it myself.

joshsizer reviewed (changes requested):
I tried simplifying the logic here, let me know what you think. One thing that still needs work is the garbage collection cronjob - if that's enabled and stateful set is enabled the cronjob doesn't mount the PVC

Nicolas-Peiffer commented on templates/_helpers.tpl:
```diff
@@ -187,7 +187,7 @@ We truncate at 63 chars because some Kubernetes name fields are limited to this
 {{- end }}
 
-{{- if eq .Values.storage "filesystem" }}
+{{- if (and (eq .Values.storage "filesystem") (not .Values.useStatefulSet)) }}
 - name: data
   {{- if .Values.persistence.enabled }}
```
I can't test this rn, but the explanation makes sense and the extra test prevent mismatch when using the `Deployment` or `StatefulSet` mode.

Nicolas-Peiffer commented on values.yaml:
```diff
@@ -63,10 +63,14 @@ resources: {}
   #  cpu: 100m
   #  memory: 128Mi
 persistence:
+  # Persistence can be achieved either with a PVC or StatefulSet. Default to PVC.
+  type: pvc
```
I do not have a strong opinion on this one.

I was inspired by this line in the [grafana `values.yaml`](https://github.com/grafana/helm-charts/blob/d5d7d3fd1c734d12675bd19a1f4a0a50294f5ede/charts/grafana/values.yaml#L417).

As long as the user of the helm chart knows without ambiguity whether he is using a Deployment or a StatfulSet, and as long as the PVC is close to each StatefulSet's pod (preferably on the same node, but this depends on which StorageClass you use), it's fine.

Nicolas-Peiffer reviewed (commented):
Hey, I had a look at your suggestions. Unfortunately, I was not able to test them, but it _looks okay_. I commented under each of your remark.

Nicolas-Peiffer committed:
Merge branch 'main' into extracontainer-stafulset-support
